Hw1

Ltaudato1 opened the pull request.

Ltaudato1 committed:
add depencies

Ltaudato1 committed:
Merge branch 'main' of https://github.com/Ltaudato1/tinkoff_edu_spring

Ltaudato1 committed:
bot base

Ltaudato1 committed:
checkstyle fix

Ltaudato1 committed:
Update .gitignore

Ltaudato1 committed:
added some tests + refactor

Ltaudato1 committed:
Update .gitignore

damir2407 commented on bot/src/main/java/edu/java/bot/BotApplication.java:
```diff
@@ -4,11 +4,18 @@
 
 @SpringBootApplication
 @EnableConfigurationProperties(ApplicationConfig.class)
+@ComponentScan("edu.java.bot")
```
Давай уберем ее? @SpringBootApplication уже включает в себе твой базовый компонент скан на модуль

damir2407 commented on bot/src/main/java/edu/java/bot/BotApplication.java:
```diff
@@ -4,11 +4,18 @@
     public static void main(String[] args) {
         SpringApplication.run(BotApplication.class, args);
+
+        BotRunner botRunner = SpringApplication
```
Тут и в других местах много вопросов про спринг, оставим их на следующие дз. В рамках этой спринг можно не менять)

damir2407 commented on .gitignore:
```diff
@@ -30,3 +31,5 @@ build/
 ### VS Code ###
 .vscode/
+bot/src/main/resources/application.yml
+*.env
```
2 раза env объявлен, определись с 1-м

damir2407 commented on .gitignore:
```diff
@@ -30,3 +31,5 @@ build/
 
 ### VS Code ###
 .vscode/
+bot/src/main/resources/application.yml
```
А зачем?) Давай уберем? Токен бота просто не стоит оставлять захардкоженным в самом коде, а можно подгружать из переменных окружения, вот так (если у тебя интелька):
<img width="797" alt="image" src="https://github.com/Ltaudato1/tinkoff_edu_spring/assets/70958074/4c28f659-0d87-43ed-ad84-df7beb339c5d">

damir2407 commented on bot/src/main/java/edu/java/bot/user/User.java:
```diff
@@ -0,0 +1,14 @@
+import java.net.URL;
+import java.util.List;
+
+public record User(long chatId, List<URL> trackedList) {
```
Имхо record'ы стоит юзать тогда, когда у тебя класс вообще не содержит какую-либо логику, а является обычной сущностью с состояниями. Я бы тут сделал обычный class. Ну, а во-вторых сложно назвать этот класс User'ом, ведь тут логика по изменению конкретных полей юзера, и будто бы это больше походит на UserManager или что-то в этом духе. Тогда ты сможешь оставить User record'ом, а методы перенести в отдельный класс UserManager

damir2407 commented on bot/src/main/java/edu/java/bot/handlers/CommandHandler.java:
```diff
@@ -0,0 +1,47 @@
+        switch (currentState) {
+            case HELP -> userMessage.helpCommand(user.chatId());
+            case START -> userMessage.startCommand(user.chatId());
+            case GET_LIST -> userMessage.listCommand(user.chatId(), user.trackedList());
```
Очень не нравится твоя логика по обслуживанию команд, как по мне выглядит не очень гибко (от слова совсем). При добавлении новой команды тебе нужно расширить аж 3 места + разрастаются switch'и до небес) Посмотри пожалуйста про паттерн Command, там можно создать интерфейс Command, с методом execute, понасоздавать команд унаследовавшись от этого интерфейса и создать класс Invoker, содержащий в себе мапу <название команды, класс реализаующий команду>. По названию команды достаешь конкретную команду и исполняешь ее execute. При таком решении все что тебе нужно - создать новую команду и добавить ее в мапу, ПРОФИТ! Давай исправим к следующей дз, в рамках этой боюсь блочить тебе выполнение 2-й)

damir2407 commented on bot/src/main/java/edu/java/bot/user/UserMessage.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            bot.execute(sendMessage);
+        } catch (Exception e) {
+            throw new Exception("failed to send message");
```
У вас есть необходимые зависимости/файлы для работы с логгированием, давай логгировать ошибки? По типу logger.error("У чувака с chatId = ** произошла такая-то ошибка"). Посмотри про аннотацию @Log4j2 от Lombok (его вам уже добавили), которая может упрастить внедрение логгера в каждый из классов

damir2407 commented on bot/src/main/resources/application.yml:
```diff
@@ -8,7 +8,6 @@ spring:
 server:
-  port: 8090
-
+  port: 0
```
Почему 0? Ощущение что у тебя этот порт должен быть занят системой, но не уверен. Давай оставим 8090?

Ltaudato1 replied:
у меня с 8090 бот не запускался, выдавало ошибку мол порт уже занят

upd: уже не актуально, исправил

damir2407 commented on bot/src/test/HandleTest.java:
```diff
@@ -0,0 +1,125 @@
+import static edu.java.bot.handlers.UserState.UNKNOWN_COMMAND;
+import static edu.java.bot.parsers.LinkParser.stringToUrl;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS) public class HandleTest {
```
Зачем тебе 1 экземпляр на весь тест?

damir2407 replied:
Посмотри на шаблон приведенный преподавателем, там в пакете test идут пакеты java/edu/java/bot. Давай не будем менять шаблон?

damir2407 commented on bot/src/test/HandleTest.java:
```diff
@@ -0,0 +1,125 @@
+import static edu.java.bot.parsers.LinkParser.stringToUrl;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS) public class HandleTest {
+    Stream<Arguments> provideData() {
```
Давай вспомогательные методы держать в самом низу класса? Чтобы повысить читаемость основной логики (в твоем случае тестов)

damir2407 commented on bot/src/test/MessageTest.java:
```diff
@@ -0,0 +1,31 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class MessageTest {
```
А че не стал добавлять тестов на корректное сообщение при различных командах?

Ltaudato1 replied:
в HandleTest есть проверки на то что вызывается корректная функция из UserMessage

damir2407 commented on pom.xml:
```diff
@@ -274,13 +294,13 @@
-                        </dependency>
-                    </dependencies>
+<!--                    <dependencies>-->
+<!--                        <dependency>-->
```
Почему закомментил? Верни пожалуйста

damir2407 commented on bot/src/main/java/edu/java/bot/MyBot.java:
```diff
@@ -0,0 +1,58 @@
+@Component
+public class MyBot {
+    @Getter private final TelegramBot bot;
+    Map<Long, User> userList;
```
Давай тоже сделаем private final? И назови как-нибудь по-другому вместо list, это же все-таки мапа. Давай например users

damir2407 replied:
А зачем тебе тогда в User хранить chatId, если у тебя он хранится в мапе? В юзере chatId даже не используется

damir2407 commented on bot/src/main/java/edu/java/bot/MyBot.java:
```diff
@@ -0,0 +1,58 @@
+public class MyBot {
+    @Getter private final TelegramBot bot;
+    Map<Long, User> userList;
+    ApplicationConfig applicationConfig;
```
Зачем тут applicationConfig? Ты же его не используешь

Ltaudato1 committed:
minor refactoring

Ltaudato1 committed:
global refactoring

Ltaudato1 committed:
hw2 is done

Ltaudato1 committed:
Update GitHubClientIntegrationTest.java

Ltaudato1 committed:
Update GitHubClientIntegrationTest.java

damir2407 commented on bot/src/main/java/edu/java/bot/handlers/ListCommand.java:
```diff
@@ -0,0 +1,17 @@
+
+    @Override
+    public void execute(Long chatId, User user, String argument) {
+        userMessage.listCommand(chatId, user.trackedList());
```
А зачем каждой команде ходить в UserMessage? Пусть каждая команда сама в себе содержит логику, просто таким образом у тебя userMessage разрастется. Можно сделать так, чтоб execute возвращал допустим String, а сам Invoker или что-то еще вызывая execute отправлял сообщение в бота.

Не нужно исправлять в этой дз, оставим на следующую

Ltaudato1 committed:
hw3 is done

Ltaudato1 committed:
update test

Ltaudato1 committed:
added menu via telegram api

Ltaudato1 committed:
hw4 is done

Ltaudato1 committed:
test fix

Ltaudato1 committed:
will be fixed in hw2

Ltaudato1 committed:
hw5 is done

Ltaudato1 committed:
hw6 is done

Ltaudato1 committed:
hw7 is done

Ltaudato1 committed:
hw8 is done

Ltaudato1 committed:
hw9

Ltaudato1 committed:
hw10

Ltaudato1 committed:
updated workflow

Ltaudato1 committed:
added jars

Ltaudato1 committed:
ACTUALLY added jars

Ltaudato1 committed:
updated docker-files

Ltaudato1 committed:
temporary fix

Ltaudato1 committed:
temporary fix v2

почему-то именно на гите докер билд вообще не проходит, хотя локально всё работает вообще отлично

Ltaudato1 committed:
refactor hw5

Ltaudato1 committed:
fix checkstyle

Ltaudato1 committed:
refactor hw7

Ltaudato1 committed:
refactor hw2

Ltaudato1 committed:
.

Ltaudato1 committed:
Merge pull request #11 from Ltaudato1/hw10

hw10

Ltaudato1 committed:
Merge pull request #10 from Ltaudato1/hw9

hw9

Ltaudato1 committed:
Merge pull request #9 from Ltaudato1/hw8

hw8 is done

Ltaudato1 committed:
Merge pull request #8 from Ltaudato1/hw7

hw7 is done

Ltaudato1 committed:
solving merge conflict pt.1

Ltaudato1 committed:
Merge branch 'hw6' of https://github.com/Ltaudato1/tinkoff_edu_spring into hw6

Ltaudato1 committed:
Merge branch 'hw6' into hw5

Ltaudato1 committed:
Merge branch 'hw3' into hw2

Ltaudato1 committed:
Merge branch 'hw2' into hw1
